File: lib/api/admin/a_announcements.ts
import { Announcement } from "@/types/announcements";
import { API_BASE_URL } from "@/lib/config";

export async function addNewAnnouncement( title: string, content: string, date: string): Promise<Announcement> {
    
  const response = await fetch (`${API_BASE_URL}/announcements/create`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      title,
      content,
      date,
    }),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Failed to create announcement: ${text}`);
  }
  return await response.json();
}

export async function editAnnouncement( id: string, title: string, content: string, date: string, archived: boolean): Promise<Announcement> {
  const response = await fetch (`${API_BASE_URL}/announcements/edit`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      id,
      title,
      content,
      date,
      archived,
    }),
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Failed to edit announcement: ${text}`);
  }
  return await response.json();
}

File: newBackend/api/admin/a_announcements.go
package apiAdmin

import (
	"backend/models"
	"backend/services"
	"encoding/json"
	"net/http"
)

// helper to apply CORS headers
func applyCORS(w http.ResponseWriter) {
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, OPTIONS")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
}

type CreateAnnouncementRequest struct {
	Title   string `json:"title"`
	Content string `json:"content"`
	Date    string `json:"date"`
}

// GET /announcements
func AddNewAnnouncement(w http.ResponseWriter, r *http.Request) {

	// allow preflight
	if r.Method == http.MethodOptions {
		w.WriteHeader(http.StatusOK)
		return
	}

	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req CreateAnnouncementRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	announcements, err := services.CreateAnnouncement(
		req.Title,
		req.Content,
		req.Date,
	)
	if err != nil {
		http.Error(w, "Failed to create announcements", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(announcements)
}

func EditAnnouncement(w http.ResponseWriter, r *http.Request) {

	// allow preflight
	if r.Method == http.MethodOptions {
		w.WriteHeader(http.StatusOK)
		return
	}

	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	var req models.Announcement
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		http.Error(w, "Invalid request body", http.StatusBadRequest)
		return
	}

	announcements, err := services.EditAnnouncement(
		req.ID,
		req.Title,
		req.Content,
		req.Date,
		req.Archived,
	)
	if err != nil {
		http.Error(w, "Failed to fetch announcements", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(announcements)
}
